executor/src/witgen/processor.rs

```diff
@@ -70,12 +73,36 @@ impl<'a, T: FieldElement> Processor<'a, T> {
         self.data[index] = self.row_factory.fresh_row();
     }
 
+    /// Figures out unknown values, using the default sequence iterator.
+    /// Since the default sequence iterator looks at the row before and after
+    /// the current block, we assume that these lines are already part of [Self::data]
+    /// and set the block size to `self.data.len() - 2`.
+    pub fn solve_with_default_sequence_iterator(&mut self) -> Result<(), EvalError<T>> {
+        let mut sequence_iterator = ProcessingSequenceIterator::Default(
+            DefaultSequenceIterator::new(self.data.len() - 2, self.identities.len()),
+        );
+        self.solve(&mut sequence_iterator)
+    }
+
     /// Figures out unknown values.
-    /// The current strategy is to go over *non-wrapping* row pairs once,
-    /// but this can be generalized in the future.
-    pub fn solve(&mut self) -> Result<(), EvalError<T>> {
-        for i in 0..(self.data.len() - 1) {
-            self.iterate_on_row_pair(i)?;
+    pub fn solve(
+        &mut self,
+        sequence_iterator: &mut ProcessingSequenceIterator,
+    ) -> Result<(), EvalError<T>> {
+        while let Some(step) = sequence_iterator.next() {
+            let SequenceStep {
+                row_delta,
+                identity,
+            } = step;
+            match identity {
+                IdentityInSequence::Internal(identity_index) => {
+                    let row_index = (1 + row_delta) as usize;
+                    let progress = self.process_identity(row_index, identity_index)?;
+                    sequence_iterator.report_progress(progress);
+                }
+                // TODO: Implement outer query
+                IdentityInSequence::OuterQuery => {}
```

Comment:
```suggestion
                IdentityInSequence::OuterQuery => { unimplemented!("Implement outer query") }
```

Reply:
Done! This actually required a change in `DefaultSequenceIterator` so that it wouldn't request an outer query, see 7d0d63bcda31dc655dffeabd0eebe0122cad911d. Think that made sense anyway.